### PR TITLE
Add decimals to gas price input

### LINKED ISSUE #1412: gasPrice does not allow numbers with decimals
I tried to set several `gasPrice` values with decimals (e.g. 2.1) and, after adding the first digit after the decimal, my number was replaced with something seemingly random.

Three examples:
2.3 --> 183
2.5 --> 185
0.3 --> 4294967279

### PR DESCRIPTION
**Pull request by 2-am-zzz**
Giving users better control over their gas! Fixes #1412 

- Allows decimals in the gasPrice input
- Fixes the reset button

### CODE REVIEW

File: ui/app/components/pending-tx.js [hunk 1]
```diff
@@ -342,22 +343,22 @@ PendingTx.prototype.miniAccountPanelForRecipient = function () {
...
-PendingTx.prototype.gasPriceChanged = function (newHex) {
-  log.info(`Gas price changed to: ${newHex}`)
-  const inWei = hexToBn(newHex).mul(GWEI_FACTOR)
+PendingTx.prototype.gasPriceChanged = function (newBN) {
+  log.info(`Gas price changed to: ${newBN.toString(10)}`)
   const txMeta = this.gatherTxMeta()
-  txMeta.txParams.gasPrice = inWei.toString(16)
-  this.setState({ txData: txMeta })
+  txMeta.txParams.gasPrice = '0x' + newBN.toString('hex')
+  this.setState({ txData: cloneObj(txMeta) })
```
Comment (kumavis) [inline ui/app/components/pending-tx.js]: why are you cloning the txMeta? were there some side effects? (if so, maybe we should do this everywhere we use `setState`)

File: ui/app/components/pending-tx.js [hunk 2]
```diff
@@ -425,14 +426,16 @@ PendingTx.prototype._notZeroOrEmptyString = function (obj) {
...
         height: '37px',
       },
     })
-
   )
 }
+
+function cloneObj (obj) {
+  return JSON.parse(JSON.stringify(obj))
+}
```
Comment (kumavis) [inline ui/app/components/pending-tx.js]: lets use the `clone` module instead https://github.com/MetaMask/metamask-plugin/blob/master/package.json#L138

**Review by kumavis — CHANGES_REQUESTED** [review]

File: ui/app/components/bn-as-decimal-input.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+BnAsDecimalInput.prototype.render = function () {
+  const props = this.props
+  const state = this.state
+
+  const { value, precision, onChange, min, max } = props
+
+  const suffix = props.suffix
+  const style = props.style
+  const scale = Math.pow(10, precision)
+  const newValue = value.toNumber(10) / scale
```
Comment (flyswatter) [inline ui/app/components/bn-as-decimal-input.js]: On this line, you are assuming that the input number fits in a JS number, which defeats the point of using BN.

I've added a test to demonstrate the problem, which highlights why this doesn't work for editing ether values, for example.

**Review by danfinlay — CHANGES_REQUESTED** [review]
I've added a test that this component should be able to pass.

**Comment by Zanibas** [comment]
@flyswatter the test you wrote is failing due to a limitation of BN.js.

`Error: Number can only safely store up to 53 bits`

Will adjust the test to accommodate this  limitation.

**Comment by flyswatter** [comment]
How is this going?

**Comment by Zanibas** [comment]
Will polish this a bit later since the functions look pretty ugly, but the logic should theoretically work.

**Comment by Zanibas** [comment]
Further fixes have been added that properly disable the submit button when invalid gas params are present.

**Review by danfinlay — APPROVED** [review]
Good job!

**Review by kumavis — APPROVED** [review]

**Pull request merged**